### PR TITLE
handle slice with negative indices and indices exceeding tensor dimen…

### PR DESCRIPTION
**Pull request by liqunfu**
handle slice with negative indices and indices exceeding tensor dimensions.

### CODE REVIEW

**Review by spandantiwari — CHANGES_REQUESTED** [review]

File: test/onnx/test_utility_funs.py [hunk 1]
```diff
@@ -48,16 +48,53 @@ def forward(self, x):
             assert node.kind() != "onnx::Constant"
         assert len(list(graph.nodes())) == 1
 
-    def test_constant_fold_slice(self):
-        class SliceModule(torch.nn.Module):
+    def test_constant_fold_narrow(self):
```
[the remark this answers is not in the document]
Comment (liqunfu) [reply test/onnx/test_utility_funs.py]: resolved

**Review by spandantiwari — APPROVED** [review]
LGTM.

File: test/onnx/test_utility_funs.py [hunk 2]
```diff
@@ -49,15 +49,52 @@ def forward(self, x):
...
+        class NarrowModule(torch.nn.Module):
             def forward(self, x):
                 a = torch.tensor([[1., 2., 3.], [4., 5., 6.]])
                 b = torch.narrow(a, 0, 0, 1)
                 return b + x
 
         _set_opset_version(9)
         x = torch.ones(1, 3)
-        graph, _, __ = utils._model_to_graph(SliceModule(), (x, ),
+        graph, _, __ = utils._model_to_graph(NarrowModule(), (x, ),
```
Comment (houseroad) [inline test/onnx/test_utility_funs.py]: nit: __ => _

also can you change all the __ to _?

File: torch/csrc/jit/passes/onnx/constant_fold.cpp [hunk 3]
```diff
@@ -87,8 +87,22 @@ c10::optional<at::Tensor> runTorchBackendForOnnx(
...
+        end = updated_val.sizes()[axis] + end;
+      }
+      // index higher than dimension is treated as the end.
+      if (end > updated_val.sizes()[axis]) {
+        end = updated_val.sizes()[axis];
+      }      
+      int64_t length = end - start;
+      if (length < 0 || start > updated_val.sizes()[axis] - length)
+        return c10::nullopt;
+      updated_val = at::narrow(updated_val, axis, start, length);
```
Comment (houseroad) [inline torch/csrc/jit/passes/onnx/constant_fold.cpp]: you may also want to try at::slice

**Review by houseroad — APPROVED** [review]
LGTM, thanks.

**Comment by houseroad** [comment]
will merge this pr first, feel free to open another pr to address nits.

**Pull request closed without merging**